src/layers/domain/core/product.py

```diff
@@ -1,98 +1,132 @@
+        self.type = type
+
+
+class RelationshipRemovedEvent(Event):
+    def __init__(self, id, target):
+        self.id = id
+        self.target = target
+
+
+class ProductKeyAddedEvent(Event):
+    def __init__(self, id, key, type):
+        self.id = id
+        self.key = key
+        self.type = type
+
 
-from .entity import Entity
-from .questionnaire_entity import QuestionnaireEntity
-from .user import User
-from .validation import (
-    validate_ods_code,
-    validate_product_id_or_asid,
-    validate_type,
-    validate_uuid,
-)
+class ProductKeyRemovedEvent(Event):
+    def __init__(self, id, key):
+        self.id = id
+        self.key = key
 
-SetType = TypeVar("SetType")
 
+class RelationshipType(Enum):
+    DEPENDENCY = "DEPENDENCY"
+    COMPONENT = "COMPONENT"
+    REVISION = "REVISION"
 
-def default_set(iterable: Iterable[SetType]) -> set[SetType]:
-    return set(iterable if iterable is not None else [])
 
+class Relationship(BaseModel):
+    type: RelationshipType = Field()
 
-class OdsOrganisation(Entity[str]):
-    """
-    Represents the bare minimum properties that we need to take from the
-    ODS definition of an organisation.
-    """
 
-    def __init__(self, id: str, name: str):
-        validate_ods_code(ods_code=id)
+class ProductKeyType(Enum):
+    PRODUCT_ID = "PRODUCT_ID"
+    ACCREDITED_SYSTEM_ID = "ACCREDITED_SYSTEM_ID"
 
-        super().__init__(id=id, name=name)
 
-    def create_product_team(
-        self, id: str, name: str, owner: User
-    ) -> tuple["ProductTeam", "ProductTeamCreatedEvent"]:
-        """
-        Creates an instance of a ProductTeam
-        """
-        product_team = ProductTeam(id=id, name=name, organisation=self, owner=owner)
-        event = ProductTeamCreatedEvent(product_team=product_team)
-        return (product_team, event)
+class ProductKey(BaseModel):
+    type: ProductKeyType = Field()
 
 
-class ProductTeam(Entity[UUID]):
-    """
-    A ProductTeam is created from a Supplier OdsOrganisation, and is the entity
-    that creates and owns Products.
-    """
+class ProductType(Enum):
+    SERVICE = "SERVICE"
+    API = "API"
 
-    def __init__(self, id: UUID, name: str, organisation: OdsOrganisation, owner: User):
-        validate_uuid(uuid=id)
-        validate_type(obj=organisation, expected_type=OdsOrganisation)
-        validate_type(obj=owner, expected_type=User)
 
-        super().__init__(id=id, name=name)
-        self.organisation = organisation
-        self.owner = owner
+class ProductStatus(Enum):
+    ACTIVE = "ACTIVE"
+    INACTIVE = "INACTIVE"
 
-    def create_product(self) -> "Product":
-        """
-        Creates an instance of a new Product
-        """
-        pass
 
+class Product(AggregateRoot):
+    id: UUID = Field()
+    name: str = Field(pattern="^[a-z][a-z0-9_]+$")
+    type: ProductType = Field()
+    status: ProductStatus = Field(default=ProductStatus.ACTIVE)
+    product_team_id: UUID = Field()
+    ods_code: str = Field()
```

Comment:
previously you had a regex for this?

Reply:
Flip flopping on whether to validate ODS codes as per elab conversation.

Reply:
yep fair - given that we can assume they're valid i think this is now moot